xknx/io/knxip_interface.py

```diff
@@ -162,3 +166,33 @@ def telegram_received(self, telegram):
     async def send_telegram(self, telegram):
         """Send telegram to connected device (either Tunneling or Routing)."""
         await self.interface.send_telegram(telegram)
+
+    def find_local_ip(self, gateway_ip):
+        """Find local IP address on same subnet as gateway."""
+        def _scan_interfaces(gateway: ipaddress.IPv4Address):
+            for interface in netifaces.interfaces():
+                try:
+                    af_inet = netifaces.ifaddresses(interface)[netifaces.AF_INET]
+                    for link in af_inet:
+                        network = ipaddress.IPv4Network((link["addr"],
+                                                         link["netmask"]),
+                                                        strict=False)
+                        if gateway in network:
+                            self.xknx.logger.debug("Using interface: %s", interface)
+                            return link["addr"]
+                except KeyError:
+                    self.xknx.logger.debug("Could not find IPv4 address on interface %s", interface)
+                    continue
+            raise XKNXException("No interface on same subnet as gateway found.")
+
+        gateway = ipaddress.IPv4Address(gateway_ip)
+        try:
```

Comment:
If above function returns `None` we could do:

```
local_ip = _scan_interfaces(gateway)
if local_ip is None:
    default_gateway = _find_default_gateway()  # <-- I would move this to a function.
    local_ip = _scan_interfaces(default_gateway)
return local_ip
```

I know that this whole block still may return `None` ... But from my understanding of networking, this should not happen under normal conditions.